reference/src/layout/unions.md

```diff
@@ -38,3 +28,44 @@ contents are.
 Even if the offsets happen to be all 0, there might still be differences in the
 function call ABI.  If you need to pass unions by-value across an FFI boundary,
 you have to use `#[repr(C)]`.
+
+### Layout of `repr(C)` unions
+
+The layout of `repr(C)` unions follows the C layout scheme. Per sections
+[6.5.8.5] and [6.7.2.1.16] of the C11 specification, this means that the offset
+of every field is 0. Unsafe code can cast a pointer to the union to a field type
+to obtain a pointer to any field, and vice versa. 
+
+[6.5.8.5]: http://port70.net/~nsz/c/c11/n1570.html#6.5.8p5
+[6.7.2.1.16]: http://port70.net/~nsz/c/c11/n1570.html#6.7.2.1p16
+
+Since all fields are at offset 0, this implies that `repr(C)` unions do not have
+padding before or in-between their fields. They can, however, have trailing
+padding (see next example).
```

Comment:
Padding isn't really part of the layout, it is part of the "value representation". So I am not sure if this fits here.

Reply:
I'm not sure either. Instead of using the word "padding" I can just say that a `union` can be larger than its largest field, and let users interpret from there that unions can have trailing padding.

Reply:
Actually the struct chapter also talks about padding. So I think I am fine mentioning it.

However, the first sentence is weird to me. Union fields are not "next to" each other, so "between" as a notion does not even make sense.

Reply:
> However, the first sentence is weird to me. Union fields are not "next to" each other, so "between" as a notion does not even make sense.

Isn't that what the sentence says? Because all fields start at offset 0, there is no space before or in-between the fields that could hold padding.

Reply:
The text talks about "in-between union fields". I think that makes the text semantically ill-formed.

I'd just remove the "or in-between", so it says

> Since all fields are at offset 0, this implies that `repr(C)` unions do not have
padding their fields.

Reply:
Well, repr(Rust) unions do not have to have all fields at offset 0, so they could have padding between fields.

Reply:
No, they can only have padding *before* their fields. Also the paragraph is (the way I read it) only about `repr(C)` unions.

Reply:
> No, they can only have padding before their fields.

Where do we guarantee this ?

Reply:
The concept "between" makes no sense, so how would they have padding there...?

"between" makes sense of structs because they put their fields into an *order* and then lay them out *next to each other*. No such thing happens for unions.

Asking for the padding between union fields IMO is like asking for the color of a union field. It's just a nonsense question. For every union field, there can be padding before it (the size of this padding equals the offset of the field) and padding after it (the size of this padding equals the size of the union minus the size of the field), and that's it.

Reply:
This snippet guarantees that, for `repr(C)` unions, all union fields have the same offset, and explicitly mentions the corollary that there cannot be any padding between union fields due to this guarantee. I think this is worth mentioning, and am open to better suggestions to explain this that do not use the word "between". If you disagree that this is worth mentioning, please argue why. 

The argument that "this is obvious"/"asking for padding is nonsense" are IMO weak, because unions do have padding, and because to the best of my knowledge we don't guarantee anywhere that all union fields have the same offset.

Reply:
I strongly feel that talking about padding "between" union fields is very confusing. I literally cannot parse that sentence, it does not have a meaning IMO.

Unions have padding, but to have something "between" two things you need to have two things *next to each other* such that there is a concept of "between". That condition is just not met here.

Reply:
I do not think anything is omitted by the sentence I proposed above:

> **Padding.** Since all fields are at offset 0, `repr(C)` unions do not have
padding before their fields. They can, however, have trailing
padding, to make sure the size is a multiple of the alignment (see next example).

---

One thing I am wondering: in a union like
```rust
#[repr(C)]
union { f1: u8, f2: u16 }
```
I feel like it makes sense to consider the first *field* as having trailing padding. Basically, all union fields are padded to the union's size. Is "padding" a good term here? Does this concept have another name?

Reply:
> I feel like it makes sense to consider the first field as having trailing padding.
> Basically, all union fields are padded to the union's size. Is "padding" a good term here? Does this concept have another name?

I think this is a good point, and I don't know if this makes sense. The first thing that jumps to mind is that the "trailing padding of a field" isn't really "padding", at least not in the same sense that the compiler isn't required to preserve its contents on move, e.g.,

```rust
#[repr(C)] union U { f1: u8, f2: u16 }
let u = U { f2: 0xff_ff };
let ptr = &u.f1 as *const _;
// Here we aren't really accessing padding:
assert_eq!(*ptr.add(1), 0xff);
```

> I strongly feel that talking about padding "between" union fields is very confusing. I literally cannot parse that sentence, it does not have a meaning IMO.
>
>Unions have padding, but to have something "between" two things you need to have two things next to each other such that there is a concept of "between". That condition is just not met here.

FWIW this is what the text attempts to explain. There is no "between" the fields, and therefore there cannot be any padding there. I am open to suggestions about better ways to put this, but I do think that it is important to call out that there is no "between" `repr(C)` union fields.

Reply:
> I think this is a good point, and I don't know if this makes sense. The first thing that jumps to mind is that the "trailing padding of a field" isn't really "padding", at least not in the same sense as for struct fields, where the compiler isn't required to preserve its contents on move

From other discussions, however, it looks like even though unions themselves can only have leading and trailing padding, but no interior padding, there can *still* be "gaps" in what is preserved by the union based on padding of the fields.

So what about describing the layout of a union as follows:

* Fields are laid out "on top of each other", which is *the* defining feature of a union compared to a struct where fields are laid out "next to each other". So we determine the size of the union (details unspecified), and then we just have to make all fields have that size by adding appropriate padding before and/or after the field. For `repr(C)`, there is no padding before the field.

    If only I could draw... in my head I have two pictures here: for a struct, if we imagine the fields as being blocks, we have to build something that looks like
    ```
    [  <--> [field 1] <--> [field 2] <--> [field 3] <--> ]
    ```
    where the `<-->` are like springs or spacers or separators -- the sizes (and alignment) of the `[field N]` blocks is fixed, the only freedom we have is to pick an order and pick the size of the gaps. These gaps are the padding.

    In contrast, for a union, this looks more like
    ```
    [ <--> [field 1] <--> ]
    [ <--> [field 2] <--> ]
    [ <--> [field 3] <--> ]
    ```
    with the extra constraint that the three "lines" here must all have the same width. So this time, the only freedom we have is to pick the padding before and after every field. See how there is no "between" that we can even talk about?

* Eventually, once we define what "the padding of a struct" and "the padding of a union" is, we can then say that a byte is a padding byte for a union if it is a padding byte for all fields. This covers both (a) trailing padding that was added to all fields, and (b) interior padding of union members that happens to "align" for all fields.

Does this make sense?

Reply:
That makes sense to me, and I do think that would be a good way to specify things.

Reply:
Feel free to use my "picture(s)". ;)

Reply:
I think that doing that is out of scope for this PR. We can open a new PR that uses the pictures for unions and structs. It something that's worth explaining at the top-level of the documents IMO.

Reply:
Fine for me -- but do you still think my proposed wording doesn't work here if we don't have the picture?

Reply:
I think it would work, but we would need to put all of it in the `repr(C)` union section only, instead of being able to use it for unions at the top level, because we don't currently guarantee anywhere (for whatever reason) that all union fields are always at the same offset. 

Most of the wording should apply to all unions anyways, so I think it would make more sense to introduce it at the top level, and only extend that with whatever is necessary in this particular section. 

Before we do that, we probably want to have consensus that we want to guarantee that all union fields for all union types start always at the same offset, since that will simplify some things. And you might want to propose that wording in the validity discussion about unions, since I think this is something that could settle the padding issue there. It's essentially the wording that others have also proposed, it would fix the `repr(transparent)` issue, making `repr(rust)` and `repr(C)` unions more consistent, etc.

Reply:
Nothing in my picture or my model has a problem with different offsets for different fields.

And I think the only reason why `repr(Rust)` does not fix offset 0 is to be able to have different offsets. For example, we unions also become subject to some amount of layout optimizations, the compiler might want to be clever here:
```rust
union Foo { b: bool, t: (u8, bool) }
```
If `b` gets offset 1 and `t` offset 0, the two `bool` align and the union could have a niche.

Reply:
That's the only reason I can imagine as well, but AFAICT those optimizations would require a field of the union to always be valid.

> Nothing in my picture or my model has a problem with different offsets for different fields.

Do we guarantee an upper bound in the size of the union, or can the fields be laid out such that there is actual padding between fields. E.g. in your `Foo` union example, `b` at offset 0, and `t` at offset 3, such that the union does have 3 actually bytes of padding between fields.

Reply:
> Do we guarantee an upper bound in the size of the union

No, not yet.

But there is still no padding between the fields. Your union looks like:
```rust
[ [b] <-- 7 bytes of padding -----------------------------> ]
[ <-- 3 bytes of padding --> [t] <-- 3 bytes of padding --> ]
```
See? No "between".

Reply:
Are you proposing that padding before and after a "field" is part of the field itself ? Because as worded before, it did not appear that way. 

So I guess what I'm missing is, what do you think a union field should be ? For a field of type `T`, just the `size_of::<T>()` bits ? or `size_of::<Union>()` bits which includes the bits of `T` ?

Reply:
No not really. I am just thinking of there being padding around every union field, as you can see in my pictures. But I think it might make sense to have something equivalent to the "variant" of an enum, where one already picked a field but did not yet "descend down into it", if that makes sense.

Notice that nowhere in my picture is there a space between two fields. Union fields provide alternative "views" onto memory. To use a union you always first pick a field, and then view the entire union according to "that field's view". If you subscribe to that view, there is no "between", not even in your example. (I sound like a broken record, don't I?).

Reply:
I guess you could say that padding is "between" two fields if some byte is in the trailing padding of one "variant" and the leading padding of another "variant". But IMO that is a complex concept, and more importantly it is *derived* from the primitive input that actually goes into union layout -- the leading and trailing padding for every "variant".

Reply:
> (I sound like a broken record, don't I?).

Not at all.

> But I think it might make sense to have something equivalent to the "variant" of an enum, where one already picked a field but did not yet "descend down into it", if that makes sense.

I think that makes sense. The padding is then in the variant, and not in the field, right ? There is definitely no padding across variants there, and because only one field is active on access, then there is only one field available each time you access the union, so it doesn't make sense to talk about padding between fields, at least not for access.

We then have bytes that aren't part of the value representation of any of the union fields. Whether they are padding before or after a field, interior field padding, etc. does not matter.

Reply:
>  but you only ever use one of these views at any given time.

Is this true ? I could pass two `&Union`  to two different threads, and each of them could access a different variant of the union simultaneously. So we might need something more precise here.

Reply:
> Is this true ? I could pass two &Union to two different threads, and each of them could access a different variant of the union simultaneously. So we might need something more precise here.

I do not think this is an issue. Multiple threads only interact on the level of untyped memory accesses ("go get me the 4 bytes starting at `0xABC00`"). The "view" is used to (a) determine which bytes to ask from memory, and (b) interpret the bytes that came back.

So there is no issue at all with thread A using view 1 and thread B using view 2, concurrently. No party involved has to consider more than one view/"variant" at a time.